File: algorithm/barkingdog/stack/BOJ17298.java
package algorithm.barkingdog.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class BOJ17298 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] inputs = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            inputs[i] = Integer.parseInt(st.nextToken());
        }

        int[] result = new int[n];
        Stack<Integer> stack = new Stack<>();
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() <= inputs[i]) {
                stack.pop();
            }
            if (stack.isEmpty()) {
                result[i] = -1;
            } else {
                result[i] = stack.peek();
            }
            stack.push(inputs[i]);
        }
        for (int i = 0; i < n; i++) {
            System.out.print(result[i] + " ");
        }
    }
}

File: algorithm/barkingdog/stack/kotlin/BOJ17298.kt
package algorithm.barkingdog.stack.kotlin

import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.util.Stack


class BOJ17298 {
}

fun main() {
    val n = readlnOrNull()?.toInt() ?: throw RuntimeException()
    val inputs = readlnOrNull()?.split(" ") ?: throw RuntimeException()
    val bw = BufferedWriter(OutputStreamWriter(System.`out`))
    val stack = Stack<Int>()
    val result = IntArray(n)
    for (i in n - 1 downTo 0) {
        val s = inputs[i].toInt()
        while (!stack.isEmpty() && stack.peek() <= s) {
            stack.pop()
        }
        if (stack.isEmpty()) {
            result[i] = -1
        } else {
            result[i] = stack.peek()
        }
        stack.push(s)
    }
    result.forEach { bw.write("$it ") }
    bw.flush()
}
